sys/senml/senml.c

```diff
@@ -0,0 +1,325 @@
+            return -1;
+        }
+    }
+
+    if (pack->base_info->base_unit) {
+        chars_written = snprintf(&output[insert_pos], len - 2, "\"bu\":\"%s\",",
+                                 pack->base_info->base_unit);
+
+        insert_pos += chars_written;
+        len -= chars_written;
+
+        if (chars_written >= (ssize_t)len - 2 || chars_written < 0) {
+            return -1;
+        }
+    }
+
+    if (pack->base_info->base_value != 0) {
+        chars_written = snprintf(&output[insert_pos], len - 2, "\"bv\":%f,",
+                                 pack->base_info->base_value);
+
+        insert_pos += chars_written;
+        len -= chars_written;
+
+        if (chars_written >= (ssize_t)len - 2 || chars_written < 0) {
+            return -1;
+        }
+    }
+
+    encode_records:
+
+    for (size_t i = 0; i < pack->num; i++) {
+        senml_record_t *curr_record = &(pack->records[i]);
+
+        if (i != 0) {
+            output[insert_pos] = '{';
+            insert_pos++;
+            len--;
+        }
+
+        if (curr_record->name) {
+            chars_written = snprintf(&output[insert_pos], len - 2, "\"n\":\"%s\",", curr_record->name);
+
+            insert_pos += chars_written;
+            len -= chars_written;
+
+            if (chars_written >= (ssize_t)len - 2 || chars_written < 0) {
+                return -1;
+            }
+        }
+
+        if (curr_record->unit) {
+            chars_written = snprintf(&output[insert_pos], len - 2, "\"u\":\"%s\",", curr_record->unit);
+
+            insert_pos += chars_written;
+            len -= chars_written;
+
+            if (chars_written >= (ssize_t)len - 2 || chars_written < 0) {
+                return -1;
+            }
+        }
+
+        if (curr_record->time != 0) {
+            chars_written = snprintf(&output[insert_pos], len - 2, "\"t\":%f,", curr_record->time);
+
+            insert_pos += chars_written;
+            len -= chars_written;
+
+            if (chars_written >= (ssize_t)len - 2 || chars_written < 0) {
+                return -1;
+            }
+        }
+
+        if (curr_record->update_time != 0) {
+            chars_written = snprintf(&output[insert_pos], len - 2, "\"ut\":%f,",
+                                     curr_record->update_time);
+
+            insert_pos += chars_written;
+            len -= chars_written;
+
+            if (chars_written >= (ssize_t)len - 2 || chars_written < 0) {
+                return -1;
+            }
+        }
+
+        if (curr_record->value_type != SENML_TYPE_UNDEF) {
+            if (curr_record->value_type == SENML_TYPE_FLOAT) {
+                chars_written = snprintf(&output[insert_pos], len - 2, "\"v\":%f",
+                                         curr_record->value.f);
+            }
+            else if (curr_record->value_type == SENML_TYPE_STRING) {
+                chars_written = snprintf(&output[insert_pos], len - 2, "\"vs\":\"%s\"",
```

Comment:
True, I'm still thinking about how to do this nicely and without wasting too much memory...